File: files/300-399/380/CH-380 Graph.py
import pandas as pd
import networkx as nx

path = "300-399/380/CH-380 Matrix Calculation.xlsx"
input_df = pd.read_excel(path, usecols="B:G", skiprows=2, nrows=5, index_col=0)
test = pd.read_excel(path, usecols="J:K", skiprows=2, nrows=5)

edges = (
    input_df.stack()
    .reset_index(name="value")
    .query("value == 1")
    [["Edge", "level_1"]]
    .rename(columns={"level_1": "to"})
)

G = nx.DiGraph()
G.add_edges_from(edges.itertuples(index=False, name=None))

result = (
    pd.DataFrame({"Factor": list(nx.topological_sort(G))})
    .reset_index(names="Rank")
    .assign(Rank=lambda df: df["Rank"] + 1)
)

print(result.equals(test))
# True

File: files/300-399/380/CH-380 Graph.R
library(tidyverse)
library(readxl)
library(igraph)

path <- "300-399/380/CH-380 Matrix Calculation.xlsx"
input <- read_excel(path, range = "B3:G8")
test <- read_excel(path, range = "J3:K8")

edges = input %>%
  pivot_longer(-Edge, names_to = "to", values_to = "value") %>%
  filter(value == 1) %>%
  transmute(from = Edge, to)

g <- graph_from_data_frame(edges, directed = TRUE)

result = topo_sort(g, mode = "out") |>
  as_ids() %>%
  data.frame(Factor = .) %>%
  mutate(Rank = row_number(), .before = 1)

all.equal(result, test, check.attributes = FALSE)
# [1] TRUE
